haskell/test/SMT/Common/Run.hs

```diff
@@ -0,0 +1,219 @@
+  nettestTestProp $
+    (uncapsNettestEmulated $ do
+        -- Set initial level
+        guardianContract <- originateSimple "guardian" () dummyGuardianContract
+        tokenContract <- originateSimple "TokenContract" [] dummyFA2Contract
+
+        -- Used in registry dao
+        registryDaoConsumer <- originateSimple "registryDaoConsumer" [] (contractConsumer @(MText, (Maybe MText)))
+
+        let PropertyTestInput (contractCalls, ms) = mkPropertyTestInput $ PropertyTestArg
+              { ptaGuardianAddr = unTAddress guardianContract
+              , ptaGovAddr = unTAddress tokenContract
+              , ptaViewContractAddr = registryDaoConsumer
+              }
+
+        advanceLevel (ms & msLevel)
+
+        -- Modifying `FullStorage` used by Registry/TreasuryDao
+        -- Update `msContracts` to track the contract called used by haskell implementation
+        let newMs = ms
+              { msFullStorage = mptoModifyFs (ms & msFullStorage)
+              , msContracts = Map.fromList
+                  [ ((unTAddress tokenContract), SimpleFA2ContractType $ SimpleFA2Contract [] (toMutez 0))
+                  , ((unTAddress registryDaoConsumer), OtherContractType $ OtherContract [] (toMutez 0))
+                  ]
+              }
+
+        -- Originate Dao for nettest
+        let
+          originateData = UntypedOriginateData
+            { uodName = "BaseDAO"
+            , uodBalance = toMutez 0
+            , uodStorage = untypeValue $ toVal $ (newMs & msFullStorage)
+            , uodContract = convertContract baseDAOContractLigo
+            }
+
+        daoUntyped <- originateUntyped originateData
+        let dao = TAddress @Parameter daoUntyped
+
+        -- The registry/treasury dao can run out of mutez
+        when (mptoContractType == RegistryDaoContract || mptoContractType == TreasuryDaoContract) $ do
+          sendXtz daoUntyped (unsafeBuildEpName "callCustom") ([mt|receive_xtz|], lPackValueRaw ())
+
+        let newMs_ = newMs { msSelfAddress = daoUntyped }
+
+        callEntrypointAndCompare (dao, tokenContract, registryDaoConsumer) contractCalls newMs_
+
+    )
+    emulatedImplIntegrational
+
+callEntrypointAndCompare :: (TAddress Parameter, TAddress FA2.Parameter, TAddress (MText, (Maybe MText))) -> [ModelCall] -> ModelState -> EmulatedT PureM ()
+callEntrypointAndCompare _ [] _ = pure ()
+callEntrypointAndCompare (dao, gov, viewC) (mc:mcs) ms = do
+
+  let (haskellErrors, updatedMs) = runBaseDaoHaskell mc ms
+      haskellStore = updatedMs & msFullStorage & fsStorage
+      govContract = updatedMs & msContracts
+        & Map.lookup (unTAddress gov)
+        & fromMaybe (error "Governance contract does not exist")
+      haskellGovStore = case govContract of SimpleFA2ContractType c -> c & sfcStorage; _ -> error "Shouldn't happen."
+      haskellGovBalance = case govContract of SimpleFA2ContractType c -> c & sfcMutez; _ -> error "Shouldn't happen."
```

Comment:
If we always need one, why is the governance contract in a map instead of having its own dedicated field ?

Reply:
The map is part of `ModelState` which track other contracts storage in our Haskell model.

Since for simplicity purpose of the generator, I used `govContract` as a fa2 contract to generate entrypoint calls that need a fa2 addr (`Transfer_contract_tokens`, `Token_transfer_type`..). 

So that's why we always need one here. This is part of the generator, and not the model, so I don't think adding a new field to the `ModelState` makes sense here. In the future, we may be interested in other contract states.